wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1054,6 +1054,102 @@ let non_null_store_assert : Env.exp_cond = fun env exp ->
     Some (Assume (BeforeExec (Constr.mk_goal "assume non-null mem store"
                                 (Bool.mk_and ctx conds))))
 
+(* For a given address, generates the constraint of:
+   addr >= RSP \/ addr <= stack_bottom - 0x256 *)
+let in_valid_mem_region (env : Env.t) (addr : Constr.z3_expr) : Constr.z3_expr =
+  (* NOTE: we are assuming stack grows down.*)
+  let ctx = Env.get_context env in
+  (* We are hardcoding the heap to be 0x256 bytes beneath the bottom of the
+     stack. *)
+  let stack_end = (Env.get_stack_end env) - 0x256 in
+  let width = env |> Env.get_sp |> Var.typ |> typ_size in
+  let sb_bv = BV.mk_numeral ctx (Int.to_string stack_end) width in
+  let stack_pointer, _ = Env.get_var env (Env.get_sp env) in
+  (* addr >= RSP *)
+  let uge = BV.mk_uge ctx addr stack_pointer in
+  (* addr <= stack_bottom - 0x256 *)
+  let ule = BV.mk_ule ctx addr sb_bv in
+  (* addr >= RSP \/ addr <= stack_bottom - 0x256 *)
+  Bool.mk_or ctx [uge; ule]
+
+(* At every memory load, add a constraint that the address is in a valid
+   location in memory. *)
+let collect_valid_mem_loads (env : Env.t) (exp : Exp.t) : Constr.z3_expr list =
```

Comment:
I notice the code for `collect_valid_mem_loads` and `collect_valid_mem_stores` only differ by one line. 
Likewise, `valid_load_vc` and `valid_store_vc` are nearly identical, and so are `valid_store_assert` and `valid_load_assert`

Should these pieces of identical-code not be abstracted? I think it looks fine as you wrote it, but I wanted to hear your (@fortunac) opinion.

Reply:
I can abstract out the similarities in `valid_load_vc` and `valid_store_vc` by creating a separate function and passing in a `collector` and the `name`, so that is no problem. As for `collect_valid_mem_loads` and `collect_valid_mem_stores`, I'm unsure how to abstract out the similarities in objects where I am overriding a method. Do you have suggestions for that?

Reply:
> I'm unsure how to abstract out the similarities in objects where I am overriding a method. Do you have suggestions for that?

I don't, and I figured that may have been an issue. Come to think of it, maybe declaring a new method should be more explicit anyway, so abstracting this out isn't really all that better.

Reply:
This is a weirdly non-modular aspect of OCaml. I wish we had a list to ask these kinds of questions.

Reply:
Okay, I pushed the changes for `valid_load_vc`, `valid_store_vc`, `valid_load_assert`, and `valid_store_assert`. I left `collect_valid_mem_loads` and `collect_valid_mem_stores`the same.